### PR TITLE
refactor secret store

### PR DESCRIPTION
**Pull request by eschultink**
### Features
  - refactor to split secret config access to a `SecretStore` interface
  - eliminate configuration potentially being in AWS Secrets Manager 
  - secret access will NEVER fall through to non-secret storage

### Change implications

 - dependencies added/changed? **no**
 - something important to note in future release notes? **idea is no ... only behavior changes in unreleased/alpha features (AWS Secrets Manager case)**

### CODE REVIEW

File: java/core/src/main/java/co/worklytics/psoxy/gateway/CompositeSecretStore.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    @NonNull
+    final SecretStore fallback;
+
+    @Override
+    public String getConfigPropertyOrError(ConfigService.ConfigProperty property) {
+        return preferred.getConfigPropertyAsOptional(property)
+            .orElseGet(() ->
+                fallback.getConfigPropertyAsOptional(property)
+                    .orElseThrow(() -> new NoSuchElementException("Missing config. no value for " + property))
+            );
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/CompositeSecretStore.java:27]:
```suggestion
        return getConfigPropertyAsOptional(property).orElseThrow(() -> new NoSuchElementException("Missing config. no value for " + property))
            );
```

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CachingConfigServiceDecorator.java [hunk 2]
```diff
@@ -17,7 +19,7 @@
 
 
 @RequiredArgsConstructor
-public class CachingConfigServiceDecorator implements ConfigService {
+public class CachingConfigServiceDecorator implements WritableConfigService, SecretStore {
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CachingConfigServiceDecorator.java:22]:
```suggestion
public class CachingConfigServiceDecorator implements SecretStore {
```
should be enough

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CompositeConfigService.java [hunk 3]
```diff
@@ -20,31 +22,14 @@
  */
 @Builder
 @RequiredArgsConstructor(access = AccessLevel.PACKAGE)
-public class CompositeConfigService implements ConfigService {
+public class CompositeConfigService implements ConfigService, SecretStore {
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CompositeConfigService.java:25]:
```suggestion
public class CompositeConfigService implements SecretStore {
```
Should be enough too

**Review by jlorper — APPROVED** [review]
Looks good to me

File: java/impl/aws/src/main/java/co/worklytics/psoxy/aws/AwsModule.java [hunk 4]
```diff
@@ -89,60 +83,84 @@ static CognitoIdentityClient cognitoClient(AwsEnvironment awsEnvironment) {
...
+            if (!vaultConfigServiceFactory.isVaultConfigured(envVarsConfigService)) {
+                throw new IllegalStateException("Vault is not configured, but HASHICORP_VAULT secret store implementation was requested");
+            }
+            VaultConfigService sharedVault =
+                vaultConfigServiceFactory.create(VaultConfigServiceFactory.pathForSharedVault(hostEnvironment, envVarsConfigService));
+
+            VaultConfigService connectorVault =
+                vaultConfigServiceFactory.create(VaultConfigServiceFactory.pathForInstanceVault(hostEnvironment, envVarsConfigService));
+
+            //initialize vaults (q: right place for this??)
```
Comment (aperez-worklytics) [inline java/impl/aws/src/main/java/co/worklytics/psoxy/aws/AwsModule.java:121]: I think the rest of the comment is clear; not ideal but the alternative is more complex and ATM we don't need it

File: java/impl/gcp/src/main/java/co/worklytics/psoxy/GcpModule.java [hunk 5]
```diff
@@ -61,6 +58,38 @@ static LockService lockService(@Named("instance") SecretManagerConfigService ins
...
+            if (!vaultConfigServiceFactory.isVaultConfigured(envVarsConfigService)) {
+                throw new IllegalStateException("Vault is not configured, but HASHICORP_VAULT secret store implementation was requested");
+            }
+            VaultConfigService sharedVault =
+                vaultConfigServiceFactory.create(VaultConfigServiceFactory.pathForSharedVault(hostEnvironment, envVarsConfigService));
+
+            VaultConfigService connectorVault =
+                vaultConfigServiceFactory.create(VaultConfigServiceFactory.pathForInstanceVault(hostEnvironment, envVarsConfigService));
+
+            //initialize vaults (q: right place for this??)
```
Comment (aperez-worklytics) [inline java/impl/gcp/src/main/java/co/worklytics/psoxy/GcpModule.java:76]: Same code than AWS, right? If that's the case, anything that we could do for reusing it? Maybe create a `VaultFactory` or something for that?

**Review by aperez-worklytics — APPROVED** [review]

File: java/impl/gcp/src/main/java/co/worklytics/psoxy/GcpModule.java (hunk 5 above)
Comment (eschultink) [reply java/impl/gcp/src/main/java/co/worklytics/psoxy/GcpModule.java]: thanks, refactored a bit.

**Pull request merged**